src/coverage/CoverageInstrumenter.ts

```diff
@@ -0,0 +1,100 @@
+import * as log4js from 'log4js';
+import { InputFile } from 'stryker-api/core';
+import { PassThrough } from 'stream';
+import { StatementMap } from 'stryker-api/test_runner';
+import { TestFramework } from 'stryker-api/test_framework';
+import { wrapInClosure } from '../utils/objectUtils';
+import CoverageInstrumenterStream from './CoverageInstrumenterStream';
+
+const log = log4js.getLogger('CoverageInstrumenter');
+
+export interface StatementMapDictionary {
+  [file: string]: StatementMap;
+}
+
+/**
+ * Represents the CoverageInstrumenter
+ * Responsible for managing the instrumentation of all files to be mutated.
+ * In case of `perTest` coverageAnalysis it will hookin to the test framework to accomplish that.
+ */
+export default class CoverageInstrumenter {
+
+  private coverageInstrumenterStreamPerFile: { [fileName: string]: CoverageInstrumenterStream } = Object.create(null);
+
+  constructor(private coverageAnalysis: 'all' | 'off' | 'perTest', private testFramework: TestFramework) { }
+
+  public instrumenterStreamForFile(file: InputFile): NodeJS.ReadWriteStream {
+    if (file.mutated) {
+      switch (this.coverageAnalysis) {
+        case 'all':
+          return this.createStreamForFile('__coverage__', file.path);
+        case 'perTest':
+          return this.createStreamForFile('__strykerCoverageCurrentTest__', file.path);
+      }
+    }
+    // By default, do not instrument for code coverage
+    return new PassThrough();
+  }
+
+  public hooksForTestRun(): string {
+    if (this.coverageAnalysis === 'perTest') {
+      log.debug(`Adding test hooks file for coverageAnalysis "perTest"`);
+      return wrapInClosure(`
+          var id = 0, coverageStateAtStart;
+          window.__coverage__ = globalCoverage = {};
+
+          ${this.testFramework.beforeEach(beforeEachFragmentPerTest)}
```

Comment:
At this point, we should always have a TestFramework. When `coverageAnalysis: 'perTest'`, stryker enforces there to be a test framework. If not, it fails fast with a message telling you how to fix it (either choose a test framework, or set coverageAnalysis to 'all'). 

I wanted to not complicate things.